File: astar_binding_src_cpp/line.cpp
#include "pch.h"

#include "Line.h"

Line::Line(Vec2 pointOnLine, Vec2 pointPerpToLine)
{
	float dx = pointOnLine.x - pointPerpToLine.x;
	float dy = pointOnLine.y - pointPerpToLine.y;

	if (dx == 0) 
	{
		m_perpendicularGradient = VERT_LINE_GRAD;
	}
	else 
	{
		m_perpendicularGradient = dy / dx;
	}

	if (m_perpendicularGradient == 0) 
	{
		m_gradient = VERT_LINE_GRAD;
	}
	else 
	{
		m_gradient = -1.0f / m_perpendicularGradient;
	}

	//this->perpendicularGradient = dx == 0 ? VERT_LINE_GRAD : dy / dx;
	//this->gradient = perpendicularGradient == 0 ? VERT_LINE_GRAD : -1.0f / perpendicularGradient;

	m_yIntercept = pointOnLine.y - m_gradient * pointOnLine.x;
	m_pointOnLine1 = pointOnLine;
	m_pointOnLine2 = pointOnLine + Vec2(1, m_gradient);

	m_approachSide = false;
	m_approachSide = GetSide(pointPerpToLine);
}

const bool Line::GetSide(Vec2 point)
{
	return (point.x - m_pointOnLine1.x) * (m_pointOnLine2.y - m_pointOnLine1.y)
				> (point.y - m_pointOnLine1.y) * (m_pointOnLine2.x - m_pointOnLine1.x);
}

const bool Line::HasCrossedLine(Vec2 point)
{
	return GetSide(point) != m_approachSide;
}

const float Line::DistanceFromPoint(Vec2 point)
{
	float yInterceptPerp = point.y - m_perpendicularGradient * point.x;
	float intersectX = (yInterceptPerp - m_yIntercept) / (m_gradient - m_perpendicularGradient);
	float intersectY = m_gradient * intersectX + m_yIntercept;
	return point.DistanceTo(Vec2(intersectX, intersectY));
}

File: _Scripts/_astar/Line.cs
using Godot;

namespace AStar
{
    /// <summary>
    /// Struct representing a line and related methods
    /// </summary>
    public struct Line
    {
        #region Constructors

        /// <summary>
        /// Initializes a new instance of the <see cref="Line"/> class.
        /// </summary>
        /// <param name="gradient">The gradient of the line</param>
        /// <param name="perpGradient">The perpendicular gradient of the line</param>
        /// <param name="point1">The point on the line</param>
        /// <param name="point2">The point perpendicular to the line</param>
        /// <param name="approachSide">The current approach side</param>
        public Line(float gradient, float perpGradient, Vector2 point1, Vector2 point2, int approachSide)
        {
            Gradient = gradient;
            PerpGradient = perpGradient;
            PointOnLine = point1;
            PointPerpToLine = point2;
            ApproachSide = approachSide == 1;
            YIntercept = point1.y - gradient * point1.x;
        }

        #endregion Constructors

        #region Properties

        /// <summary>
        /// Gets the gradient of the line.
        /// </summary>
        public float Gradient { get; }

        /// <summary>
        /// Gets the y intercept of the line.
        /// </summary>
        public float YIntercept { get; }

        /// <summary>
        /// Gets the perpendicular gradient of the line.
        /// </summary>
        public float PerpGradient { get; }

        /// <summary>
        /// Gets the point on the line.
        /// </summary>
        public Vector2 PointOnLine { get; }

        /// <summary>
        /// Gets the point perpendicular to the line.
        /// </summary>
        public Vector2 PointPerpToLine { get; }

        /// <summary>
        /// Gets the approach side, where true represents front
        /// and false back.
        /// </summary>
        public bool ApproachSide { get; }

        #endregion Properties

        #region Public Methods

        /// <summary>
        /// Getter method that gets the side of the passed point.
        /// </summary>
        /// <param name="point">The passed point</param>
        /// <returns>The side the point represented as a bool (infront or behind)</returns>
        public bool GetSide(Vector2 point)
        {
            return (point.x - PointOnLine.x) * (PointPerpToLine.y - PointOnLine.y) > (point.y - PointOnLine.y) * (PointPerpToLine.x - PointOnLine.x);
        }

        /// <summary>
        /// Determines whether the passed point has crossed the line.
        /// </summary>
        /// <param name="point">The passed point</param>
        /// <returns>boolean of whether the point has crossed the line</returns>
        public bool HasCrossedLine(Vector2 point)
        {
            return GetSide(point) != ApproachSide;
        }

        /// <summary>
        /// Gets the distance of the line to the passed point.
        /// </summary>
        /// <param name="point">The passed point</param>
        /// <returns>The distance of the line to the point</returns>
        public float DistanceFromPoint(Vector2 point)
        {
            float yInterceptPerp = point.y - PerpGradient * point.x;
            float xIntersect = (yInterceptPerp - YIntercept) / (Gradient - PerpGradient);
            float yIntersect = Gradient * xIntersect + YIntercept;
            return point.DistanceTo(new Vector2(xIntersect, yIntersect));
        }

        #endregion Public Methods
    }
}